attributecert/attributecert.go

```diff
@@ -0,0 +1,713 @@
+	ComponentIdentifiers    []ComponentIdentifierV2 `asn1:"optional,tag=0"`
+	ComponentIdentifiersUri UriReference            `asn1:"optional,tag=1"`
+	PlatformProperties      []Property              `asn1:"optional,tag=2"`
+	PlatformPropertiesUri   UriReference            `asn1:"optional,tag=3"`
+}
+
+type ComponentIdentifierV1 struct {
+	ComponentManufacturer   string
+	ComponentModel          string
+	ComponentSerial         string             `asn1:"optional,tag=0"`
+	ComponentRevision       string             `asn1:"optional,tag=1"`
+	ComponentManufacturerId int                `asn1:"optional,tag=2"`
+	FieldReplaceable        bool               `asn1:"optional,tag=3"`
+	ComponentAddresses      []ComponentAddress `asn1:"optional,tag=4"`
+}
+
+type PlatformConfigurationV1 struct {
+	ComponentIdentifiers  []ComponentIdentifierV1 `asn1:"optional,tag=0"`
+	PlatformProperties    []Property              `asn1:"optional,tag=1"`
+	PlatformPropertiesUri []UriReference          `asn1:"optional,tag=2"`
+}
+
+func parseAttributeCertificate(in *attributeCertificate) (*AttributeCertificate, error) {
+	out := new(AttributeCertificate)
+	out.Raw = in.Raw
+	out.RawTBSAttributeCertificate = in.TBSAttributeCertificate.Raw
+
+	out.Signature = in.SignatureValue.RightAlign()
+	out.SignatureAlgorithm = getSignatureAlgorithmFromAI(in.TBSAttributeCertificate.SignatureAlgorithm)
+	out.Version = in.TBSAttributeCertificate.Version + 1
+	out.SerialNumber = in.TBSAttributeCertificate.SerialNumber
+
+	var v asn1.RawValue
+	_, err := asn1.Unmarshal(in.TBSAttributeCertificate.Issuer.IssuerName.Bytes, &v)
+	if err != nil {
+		return nil, err
+	}
+
+	var issuer pkix.RDNSequence
+	if rest, err := asn1.Unmarshal(v.Bytes, &issuer); err != nil {
+		return nil, err
+	} else if len(rest) != 0 {
+		return nil, errors.New("x509: trailing data after X.509 subject")
+	}
+
+	out.Issuer.FillFromRDNSequence(&issuer)
+
+	_, err = asn1.Unmarshal(in.TBSAttributeCertificate.Holder.BaseCertificateID.Issuer.Bytes, &v)
+	if err != nil {
+		return nil, err
+	}
+
+	var holder pkix.RDNSequence
+	if rest, err := asn1.Unmarshal(v.Bytes, &holder); err != nil {
+		return nil, err
+	} else if len(rest) != 0 {
+		return nil, errors.New("x509: trailing data after X.509 subject")
+	}
+
+	out.Holder.Issuer.FillFromRDNSequence(&holder)
+	out.Holder.Serial = in.TBSAttributeCertificate.Holder.BaseCertificateID.Serial
+
+	out.NotBefore = in.TBSAttributeCertificate.Validity.NotBefore
+	out.NotAfter = in.TBSAttributeCertificate.Validity.NotAfter
+
+	var attributes []attribute
+	for _, rawAttr := range in.TBSAttributeCertificate.RawAttributes {
+		var parsedAttribute attribute
+		rest, err := asn1.Unmarshal(rawAttr.FullBytes, &parsedAttribute)
+		if err == nil && len(rest) == 0 {
+			attributes = append(attributes, parsedAttribute)
+		} else {
+			return nil, err
+		}
+	}
+
+	for _, attribute := range attributes {
+		if attribute.Id.Equal(oidAttributeUserNotice) {
+			for _, value := range attribute.RawValues {
```

Comment:
If this is a scalar field, whats the intent of using a for loop over the `RawValues` field, if only the last one will win? what am I missing?

Reply:
It's defined as a set of values, but given we're not using that the loop is unnecessary.